scripting/src/main/java/org/infinispan/scripting/impl/DataType.java

```diff
@@ -0,0 +1,105 @@
+package org.infinispan.scripting.impl;
+
+import org.infinispan.commons.CacheException;
+import org.infinispan.commons.marshall.Marshaller;
+import org.infinispan.commons.util.Immutables;
+
+import java.io.UnsupportedEncodingException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public enum DataType {
+   UTF8(new Utf8Transformer()),
+   DEFAULT(new DefaultTransformer());
+
+   public final Transformer transformer;
+
+   DataType(Transformer transformer) {
+      this.transformer = transformer;
+   }
+
+   public static DataType fromMime(String mime) {
+      switch (mime) {
+         case "text/plain; charset=utf-8":
+            return UTF8;
+         default:
+            return DEFAULT;
+      }
+   }
+
+   interface Transformer {
+      Map<String, ?> toDataType(Map<String, ?> objs, Optional<Marshaller> marshaller);
+      Object fromDataType(Object obj, Optional<Marshaller> marshaller);
+   }
+
+   static final class Utf8Transformer implements Transformer {
+      @Override
+      public Map<String, ?> toDataType(Map<String, ?> objs, Optional<Marshaller> marshaller) {
+         return objs.entrySet().stream().map(e -> {
+            Object v = e.getValue();
+            Object entryValue = v instanceof byte[] ? asString(v) : v;
+            return Immutables.immutableEntry(e.getKey(), entryValue);
+         }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+      }
+
+      @Override
+      public Object fromDataType(Object obj, Optional<Marshaller> marshaller) {
+         try {
+            return obj instanceof String ? ((String) obj).getBytes("utf8") : obj;
+         } catch (UnsupportedEncodingException e) {
+            throw new CacheException(e);
+         }
+      }
+
+      private static String asString(Object v) {
+         try {
+            return new String((byte[]) v, "utf8");
```

Comment:
Use the new String(null, Charset.forName("UTF-8")); form to avoid the try/catch. Constify the UTF-8 Charset elsewhere for extra points ;)